File: extension/lib/protocol.js
// extension/lib/protocol.js
//
// Single source of truth for the bridge wire protocol.
// Loaded as an ES module by background.js, popup.js, sandbox.js.
// Mirrored by worker-template/src/types.ts and python-template/onboard_common/protocol.py.
//
// Protocol version is bumped on any breaking change to command shapes or
// message envelopes. The extension advertises its version in the `connect`
// message; the server MUST refuse mismatched versions with a clear error.

export const PROTOCOL_VERSION = "1.0";

// ---------------------------------------------------------------------------
// Server -> Extension messages
// ---------------------------------------------------------------------------

export const CMD = Object.freeze({
  // Heartbeat
  PING: "ping",

  // Fetch family — execute fetch() in different contexts
  FETCH: "fetch",              // service-worker context (handles zstd via DecompressionStream)
  PAGE_FETCH: "page.fetch",    // page main-world context (native zstd, required for Chrome 151+)

  // Tab family
  TABS_OPEN: "tabs.open",
  TABS_CLOSE: "tabs.close",
  TABS_LIST: "tabs.list",
  TABS_FOCUS: "tabs.focus",

  // Form family — DOM interaction in a tab
  FORM_FILL: "form.fill",      // React-safe input value setter
  FORM_CLICK: "form.click",
  FORM_WAIT: "form.wait",      // poll for selector to appear
  FORM_EVAL: "form.eval",      // CDP Runtime.evaluate in page main world (bypasses CSP)

  // Network capture
  XHR_INTERCEPT: "xhr.intercept",  // one-shot capture of a single matching request

  // Cookies
  COOKIES_GET: "cookies.get",
  COOKIES_GET_ALL: "cookies.getAll",
  COOKIES_SET: "cookies.set",

  // Visual capture
  SCREENSHOT: "screenshot",          // visible tab (tabId-aware, fixes legacy bug)

  // Captcha — multi-provider
  CAPTCHA_GET_TOKEN: "captcha.getToken",  // args.provider: hcaptcha|recaptcha|turnstile|cloudflare

  // Sandbox
  SANDBOX_OPEN: "sandbox.open",
  SANDBOX_FETCH: "sandbox.fetch",    // routed to sandbox page (page-context fetch, zstd-native)

  // Activatable debugging (toggle on/off per tab)
  DEBUG_HAR_START: "debug.har.start",
  DEBUG_HAR_STOP: "debug.har.stop",
  DEBUG_CONSOLE_START: "debug.console.start",
  DEBUG_CONSOLE_STOP: "debug.console.stop",
  DEBUG_NETWORK_START: "debug.network.start",
  DEBUG_NETWORK_STOP: "debug.network.stop",
  DEBUG_TRACE_START: "debug.trace.start",
  DEBUG_TRACE_STOP: "debug.trace.stop",
  DEBUG_DOM_SNAPSHOT: "debug.dom.snapshot",
  DEBUG_STORAGE_DUMP: "debug.storage.dump",
  DEBUG_SCREENSHOT_FULL: "debug.screenshot.fullpage",
});

// All known command types — used for capability advertisement.
export const ALL_COMMAND_TYPES = Object.freeze(Object.values(CMD));

// ---------------------------------------------------------------------------
// Extension -> Server messages
// ---------------------------------------------------------------------------

export const MSG = Object.freeze({
  AUTH: "auth",          // {type, token} — sent immediately on WS open
  CONNECT: "connect",    // {type, agentId, protocolVersion, capabilities, context, userAgent, hostname}
  AUTH_OK: "auth-ok",    // server -> ext only
  RESULT: "result",      // {type, id, ok, ...payload} — synchronous command result
  LOG: "log",            // {type, level, message, data, commandId?, tabId?, durationMs?}
  PONG: "pong",          // {type, ts} — keepalive reply (also used as unsolicited keepalive)
  EVENT: "event",        // {type, event, tabId, data} — async event from activatable debug streams
  STATUS: "status",      // {type, status} — periodic status broadcast (popup consumes)
});

// ---------------------------------------------------------------------------
// Log levels
// ---------------------------------------------------------------------------

export const LOG_LEVEL = Object.freeze({
  DEBUG: "debug",
  INFO: "info",
  WARN: "warn",
  ERROR: "error",
});

// ---------------------------------------------------------------------------
// Connection contexts — extension vs sandbox page
// ---------------------------------------------------------------------------

export const CONTEXT = Object.freeze({
  WORKER: "worker",   // background service worker (default)
  PAGE: "page",       // sandbox.html — page-context for zstd-native fetch
});

// ---------------------------------------------------------------------------
// Capability advertisement
//
// The extension sends this list in the `connect` message so the server can
// probe before sending commands that require optional permissions (e.g.
// `debugger` for CDP-based commands). Future extensions can advertise
// additional capabilities without changing the protocol.
// ---------------------------------------------------------------------------

export function buildCapabilities() {
  return ALL_COMMAND_TYPES.filter((t) => t !== "ping");
}

// ---------------------------------------------------------------------------
// Command envelope validation
//
// Every command from the server MUST have:
//   - type: string, one of CMD.*
//   - id:   string, server-generated correlation ID
// Optional fields:
//   - protocolVersion: string (defaults to "1.0" if absent for back-compat)
//   - traceId: string, server-generated end-to-end trace ID (flows to logs/results)
//   - timeoutMs: number, per-command timeout
// ---------------------------------------------------------------------------

export function validateCommand(msg) {
  if (!msg || typeof msg !== "object") return "command is not an object";
  if (typeof msg.type !== "string") return "command.type is not a string";
  if (!ALL_COMMAND_TYPES.includes(msg.type)) return `unknown command type: ${msg.type}`;
  if (typeof msg.id !== "string" || msg.id.length === 0) return "command.id is missing or empty";
  return null;
}

// ---------------------------------------------------------------------------
// Default timeouts (ms) — per command type
// ---------------------------------------------------------------------------

export const DEFAULT_TIMEOUT_MS = Object.freeze({
  [CMD.FETCH]: 30_000,
  [CMD.PAGE_FETCH]: 30_000,
  [CMD.TABS_OPEN]: 60_000,
  [CMD.TABS_CLOSE]: 5_000,
  [CMD.TABS_LIST]: 5_000,
  [CMD.TABS_FOCUS]: 5_000,
  [CMD.FORM_FILL]: 10_000,
  [CMD.FORM_CLICK]: 10_000,
  [CMD.FORM_WAIT]: 30_000,
  [CMD.FORM_EVAL]: 30_000,
  [CMD.XHR_INTERCEPT]: 30_000,
  [CMD.COOKIES_GET]: 5_000,
  [CMD.COOKIES_GET_ALL]: 5_000,
  [CMD.COOKIES_SET]: 5_000,
  [CMD.SCREENSHOT]: 10_000,
  [CMD.CAPTCHA_GET_TOKEN]: 15_000,
  [CMD.SANDBOX_OPEN]: 5_000,
  [CMD.SANDBOX_FETCH]: 120_000,
  [CMD.DEBUG_HAR_START]: 5_000,
  [CMD.DEBUG_HAR_STOP]: 10_000,
  [CMD.DEBUG_CONSOLE_START]: 5_000,
  [CMD.DEBUG_CONSOLE_STOP]: 10_000,
  [CMD.DEBUG_NETWORK_START]: 5_000,
  [CMD.DEBUG_NETWORK_STOP]: 10_000,
  [CMD.DEBUG_TRACE_START]: 5_000,
  [CMD.DEBUG_TRACE_STOP]: 30_000,
  [CMD.DEBUG_DOM_SNAPSHOT]: 15_000,
  [CMD.DEBUG_STORAGE_DUMP]: 10_000,
  [CMD.DEBUG_SCREENSHOT_FULL]: 15_000,
});

export function timeoutFor(msg) {
  if (typeof msg.timeoutMs === "number" && msg.timeoutMs > 0) return msg.timeoutMs;
  return DEFAULT_TIMEOUT_MS[msg.type] ?? 30_000;
}

File: python-template/onboard_common/protocol.py
"""onboard_common.protocol

Python mirror of the wire protocol. Single source of truth for the Python
side. Keep in sync with extension/lib/protocol.js and worker-template/src/types.ts.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Optional
import uuid

PROTOCOL_VERSION = "1.0"


class CommandType:
    PING = "ping"
    FETCH = "fetch"
    PAGE_FETCH = "page.fetch"
    TABS_OPEN = "tabs.open"
    TABS_CLOSE = "tabs.close"
    TABS_LIST = "tabs.list"
    TABS_FOCUS = "tabs.focus"
    FORM_FILL = "form.fill"
    FORM_CLICK = "form.click"
    FORM_WAIT = "form.wait"
    FORM_EVAL = "form.eval"
    XHR_INTERCEPT = "xhr.intercept"
    COOKIES_GET = "cookies.get"
    COOKIES_GET_ALL = "cookies.getAll"
    COOKIES_SET = "cookies.set"
    SCREENSHOT = "screenshot"
    CAPTCHA_GET_TOKEN = "captcha.getToken"
    SANDBOX_OPEN = "sandbox.open"
    SANDBOX_FETCH = "sandbox.fetch"
    DEBUG_HAR_START = "debug.har.start"
    DEBUG_HAR_STOP = "debug.har.stop"
    DEBUG_CONSOLE_START = "debug.console.start"
    DEBUG_CONSOLE_STOP = "debug.console.stop"
    DEBUG_NETWORK_START = "debug.network.start"
    DEBUG_NETWORK_STOP = "debug.network.stop"
    DEBUG_TRACE_START = "debug.trace.start"
    DEBUG_TRACE_STOP = "debug.trace.stop"
    DEBUG_DOM_SNAPSHOT = "debug.dom.snapshot"
    DEBUG_STORAGE_DUMP = "debug.storage.dump"
    DEBUG_SCREENSHOT_FULL = "debug.screenshot.fullpage"


class MessageType:
    AUTH = "auth"
    CONNECT = "connect"
    AUTH_OK = "auth-ok"
    RESULT = "result"
    LOG = "log"
    PONG = "pong"
    EVENT = "event"
    STATUS = "status"


class LogLevel:
    DEBUG = "debug"
    INFO = "info"
    WARN = "warn"
    ERROR = "error"


class Context:
    WORKER = "worker"
    PAGE = "page"


# Default timeouts (ms) — must match extension/lib/protocol.js
DEFAULT_TIMEOUT_MS = {
    CommandType.FETCH: 30_000,
    CommandType.PAGE_FETCH: 30_000,
    CommandType.TABS_OPEN: 60_000,
    CommandType.TABS_CLOSE: 5_000,
    CommandType.TABS_LIST: 5_000,
    CommandType.TABS_FOCUS: 5_000,
    CommandType.FORM_FILL: 10_000,
    CommandType.FORM_CLICK: 10_000,
    CommandType.FORM_WAIT: 30_000,
    CommandType.FORM_EVAL: 30_000,
    CommandType.XHR_INTERCEPT: 30_000,
    CommandType.COOKIES_GET: 5_000,
    CommandType.COOKIES_GET_ALL: 5_000,
    CommandType.COOKIES_SET: 5_000,
    CommandType.SCREENSHOT: 10_000,
    CommandType.CAPTCHA_GET_TOKEN: 15_000,
    CommandType.SANDBOX_OPEN: 5_000,
    CommandType.SANDBOX_FETCH: 120_000,
    CommandType.DEBUG_HAR_START: 5_000,
    CommandType.DEBUG_HAR_STOP: 10_000,
    CommandType.DEBUG_CONSOLE_START: 5_000,
    CommandType.DEBUG_CONSOLE_STOP: 10_000,
    CommandType.DEBUG_NETWORK_START: 5_000,
    CommandType.DEBUG_NETWORK_STOP: 10_000,
    CommandType.DEBUG_TRACE_START: 5_000,
    CommandType.DEBUG_TRACE_STOP: 30_000,
    CommandType.DEBUG_DOM_SNAPSHOT: 15_000,
    CommandType.DEBUG_STORAGE_DUMP: 10_000,
    CommandType.DEBUG_SCREENSHOT_FULL: 15_000,
}


@dataclass
class Command:
    """A command sent to the extension."""
    type: str
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    trace_id: Optional[str] = None
    timeout_ms: Optional[int] = None
    payload: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        d = {"type": self.type, "id": self.id, **self.payload}
        if self.trace_id:
            d["traceId"] = self.trace_id
        if self.timeout_ms:
            d["timeoutMs"] = self.timeout_ms
        d["protocolVersion"] = PROTOCOL_VERSION
        return d


@dataclass
class CommandResult:
    """Result of executing a command on the extension."""
    ok: bool
    data: dict[str, Any] = field(default_factory=dict)
    error: Optional[str] = None
    trace_id: Optional[str] = None
    duration_ms: Optional[int] = None

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "CommandResult":
        return cls(
            ok=bool(d.get("ok", False)),
            data={k: v for k, v in d.items() if k not in ("type", "id", "ok", "error", "traceId", "durationMs")},
            error=d.get("error"),
            trace_id=d.get("traceId"),
            duration_ms=d.get("durationMs"),
        )

    def get(self, key: str, default: Any = None) -> Any:
        return self.data.get(key, default)

    @property
    def status(self) -> Optional[int]:
        return self.data.get("status")

    @property
    def body(self) -> Optional[str]:
        return self.data.get("body")

    @property
    def headers(self) -> Optional[dict]:
        return self.data.get("headers")

    @property
    def final_url(self) -> Optional[str]:
        return self.data.get("finalUrl")

    def raise_for_error(self) -> None:
        """Raise an exception if the command failed."""
        if not self.ok:
            from .exceptions import BridgeCommandError
            raise BridgeCommandError(self.error or "Command failed", command_result=self)
